### DISCUSSION TITLE
Update to Documentation on WASM Support

### DISCUSSION THREAD

**Discussion by stellarpower** [containers/podman#17115] [General]
Hi, trying to run my first WASM container with podman.

I've followed tutorials like this one:
[OCI containers with Wasm](https://www.sobyte.net/post/2022-12/oci-wasm/)
Built crun from source with wasmedge enabled:
```
$ crun --version
crun version 1.7.2.0.0.0.61-e7b94
commit: e7b94c4c512613b5be6b3aafbef5d698821f9d28
rundir: /run/user/1000/crun
spec: 1.0.0
+SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +WASM:wasmedge +YAJL
```
I still get an exec format error when I try to run a container:
`podman run --rm --annotation module.wasm.image/variant=compat-smart docker.io/wasmedge/example-wasi:latest /wasi_example_main.wasm 50000000`

I presume crun should be the default preferred runtime. There are some steps on [this page](https://github.com/containers/crun/blob/main/docs/wasm-wasi-example.md) that I don't understand without reference to the files involved or more details.

Am I missing something? Is there a minimum podman version that is needed for WASM support? I presume I will need to configure something somewhere to handle the annotation, but that's what I don't see detailed online in general. If the CRUN docs could get updated that might be helpful.

Thanks

**Comment by rhatdan** [comment]
@flouthoc @giuseppe PTAL

**Comment by giuseppe** [comment]
a new enough podman would do it automatically but if you don't have it, you can use `--annotation run.oci.handler=wasm` instead of `--annotation module.wasm.image/variant=compat-smart`

**Comment by stellarpower** [comment]
Hmmm...
`Error: OCI runtime error: invalid handler specified 'wasm': Invalid argument`

**Reply by giuseppe** [reply]
do you get the same error with `--annotation run.oci.handler=wasmedge`?

Can you make sure podman is using the new crun you've built?

**Reply by stellarpower** [reply]
I get the same, although the wasmedge installer doesn't seem to work for me - the `curl | bash` script form their site runs to completion, and then I don't have it installed in the path or any trace in the filesystem, so that's more fun :) I did try with `wasmtime` , which is installed though

Is there a way I can test with the crun version? I'll `strace` it or something for now.

**Reply by stellarpower** [reply]
Okay, wasmedge installer still fails, but I managed to find where it dumped the files. I forced the lib directory into the LD_LIBRARY_PATH (although`ldd` says it's using a relative RPAth and correctly finds it and runs as a standalone):
```
$ podman run --runtime /usr/local/bin/crun  --rm   --annotation run.oci.handler=wasmedge   localhost/mywasm-image
Error: could not load `libwasmedge.so.0`: libwasmedge.so.0: cannot open shared object file: No such file or directory: OCI runtime attempted to invoke a command that was not found
```
Also temporarily symlinked into /usr/local/lib just in case ENV variables are getting nuked at this stage, but ot no avail.
Same for the image from before. Thanks for your time on this BTW

**Reply by giuseppe** [reply]
is there a `wasmedge` binary installed as well?  Does that work if you call it directly?

You may need to install the libraries under `/usr/lib` instead of `/usr/local`

**Reply by stellarpower** [reply]
Thanks. It seems it's not looking into `/usr/local/lib`, even though it's in `/etc/ld.conf` and even if I force into `LD_LIBRARY_PATH`. That's where my patched crun also lives, interestingly; as on the shell the newer version correctly is found and takes precedence over the old from the package manager.

If there's a way to fix that and it'd be appropriate, I think that would be preferable, as I'd rather not link into `/usr/lib`, which should really be fully under the control of the package manager, but can confirm it works doing this for now:

```
$ podman run  --runtime /usr/local/bin/crun    mywasm-image:latest

Hello, world!
$ podman run  --runtime /usr/local/bin/crun    docker.io/wasmedge/example-wasi:latest /wasi_example_main.wasm 50000000

Random number: -1909555752
Random bytes: [38, 225, 96, 216, 101, 110, 98, 105, 70, 77, 109, 217, 138, 102, 17, 127, 75, 183, 214, 227, 125, 117, 140, 20, 38, 101, 79, 91, 169, 64, 150, 59, 10, 12, 253, 114, 159, 50, 197, 168, 171, 186, 135, 11, 85, 151, 203, 240, 225, 102, 158, 33, 191, 210, 77, 132, 4, 223, 128, 97, 140, 215, 27, 4, 198, 216, 171, 14, 71, 249, 230, 24, 167, 235, 248, 61, 34, 110, 138, 130, 120, 183, 34, 72, 134, 218, 58, 158, 86, 32, 30, 9, 185, 163, 76, 188, 49, 222, 78, 8, 92, 0, 31, 95, 97, 231, 225, 239, 179, 169, 158, 212, 182, 98, 153, 152, 106, 166, 160, 24, 251, 118, 136, 66, 161, 61, 44, 211]
Printed from wasi: This is from a main function
This is from a main function
The env vars are as follows.
PATH: /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
TERM: xterm
container: podman
HOSTNAME: 8e4aaada6c5b
HOME: /
The args are as follows.
/wasi_example_main.wasm
50000000
File content is This is in a file
$ 
```

So I don't even need the annotation here (although I guess it still features in the image); with the wasmedge library found, via some mechanism it's making the decision.

Thanks for your help! I'll try to spin some of this out into an explanation and bulk up the `crun` documentation page, if I find time.

**Comment by stellarpower** [comment]
`podman version 3.4.2` BTW, think I didn't mention already. We're stuck on focal for the time being; do you happen to know of an upgrade path to P4? I know it requires doing a `system reset` so it won;t be an option right now, but I'd like ot move on if we can.

**Reply by Luap99** [reply]
>  I know it requires doing a system reset so it won;t be an option right now, but I'd like ot move on if we can.

It is not required, not sure why you think so. You can just keep using it like that and it will keep using CNI.

**Reply by stellarpower** [reply]
That is what we had to do when we upgraded from 3 to 4 previously, about a year ago.
